client/commands.c

```diff
@@ -48,73 +50,221 @@ void processXferCommand(char *nextParam, FILE *out, int ispwd, uint32_t* pwd);
+}
+
+int xdag_com_block(char * args, FILE* out) {
+    processBlockCommand(args, out);
+    return 0;
+}
+
+int xdag_com_lastblocks(char * args, FILE* out) {
+    processLastBlocksCommand(args, out);
+    return 0;
+}
+
+int xdag_com_keyGen(char * args, FILE* out) {
+    processKeyGenCommand(out);
+    return 0;
+}
+
+int xdag_com_level(char * args, FILE* out) {
+    processLevelCommand(args, out);
+    return 0;
+}
+
+int xdag_com_mining(char * args, FILE* out) {
+    processMiningCommand(args, out);
+    return 0;
+}
+
+int xdag_com_net(char * args, FILE* out) {
+    processNetCommand(args, out);
+    return 0;
+}
+
+int xdag_com_pool(char * args, FILE* out) {
+    processPoolCommand(args, out);
+    return 0;
+}
+
+int xdag_com_miners(char * args, FILE* out) {
+    processMinersCommand(args, out);
+    return 0;
+}
+
+int xdag_com_stats(char * args, FILE* out) {
+    processStatsCommand(out);
+    return 0;
+}
+
+int xdag_com_state(char * args, FILE* out) {
+    fprintf(out, "%s\n", get_state());
+    return 0;
+}
+
+int xdag_com_run(char * args, FILE* out) {
+    g_xdag_run = 1;
+    return 0;
+}
+
+int xdag_com_terminate(char * args, FILE* out) {
+    processExitCommand();
+    return -1;
+}
+
+int xdag_com_exit(char * args, FILE* out) {
+    processExitCommand();
+    return -1;
+}
+
+int xdag_com_help(char *args, FILE* out) {
+    printHelp(out);
+    return 0;
+}
+
+char ** xdag_com_completion(const char *text, int start, int end) {
+    char **matches;
+    matches = (char **)NULL;
+    if (start == 0)
+        matches = rl_completion_matches(text, xdag_com_generator);
+    return (matches);
+}
+
+char* dupstr(char* s){
+    char *r;
+    r = malloc(strlen(s) + 1);
+    strcpy(r, s);
+    return (r);
+}
+
+char* xdag_com_generator(const char* text, int state) {
+    static int list_index, len;
+    char *name;
+    if (!state) {
+        list_index = 0;
+        len = strlen(text);
+    }
+    
+    while ((name = commands[list_index].name)) {
+        list_index++;
+        if (strncmp(name, text, len) == 0)
+            return (dupstr(name));
+    }
+    return ((char *)NULL);
+}
+
+XDAG_COMMAND* find_xdag_command(char *name) {
+    register int i;
+    for (i = 0; commands[i].name; i++)
+        if (strcmp(name, commands[i].name) == 0)
+            return (&commands[i]);
+    return ((XDAG_COMMAND *)NULL);
+}
+
 void startCommandProcessing(int transportFlags)
 {
-	char cmd[XDAG_COMMAND_MAX] = {0};
-
-	if(!(transportFlags & XDAG_DAEMON)) printf("Type command, help for example.\n");
-	for(;;) {
-		if(transportFlags & XDAG_DAEMON) sleep(100);
-		else {
-			printf("%s> ", g_progname);
-			fflush(stdout);
-			fgets(cmd, XDAG_COMMAND_MAX, stdin);
-			if(xdag_command(cmd, stdout) < 0) {
-				break;
-			}
-		}
-	}
+    char *cmd;
+    if(!(transportFlags & XDAG_DAEMON)) printf("Type command, help for example.\n");
+    
+    // readline lib
+    rl_readline_name = "xdag";
+    rl_attempted_completion_function = xdag_com_completion;
```

Comment:
xdag_com_completion mallocs memory, will it be freed in readline lib or manually?

Reply:
i check readline lib complete.c :
1160   if (rl_attempted_completion_function)
1161     {
1162       matches = (*rl_attempted_completion_function) (text, start, end);
1163       if (RL_SIG_RECEIVED())
1164         {
1165           _rl_free_match_list (matches);
1166           matches = 0;
1167           RL_CHECK_SIGNALS ();
1168         }
1169 
1170       if (matches || rl_attempted_completion_over)
1171         {
1172           rl_attempted_completion_over = 0;
1173           return (matches);
1174         }
1175     }

_rl_free_match_list() while free malloc memory 

1897 void
1898 _rl_free_match_list (matches)
1899      char **matches;
1900 {
1901   register int i;
1902 
1903   if (matches == 0)
1904     return;
1905 
1906   for (i = 0; matches[i]; i++)
1907     xfree (matches[i]);
1908   xfree (matches);
1909 }